flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/CreditBasedPartitionRequestClientHandlerTest.java

```diff
@@ -419,6 +456,68 @@ public void testNotifyCreditAvailableAfterReleased() throws Exception {
 		}
 	}
 
+	@Test
+	public void testReceivedBufferForRemovedChannel() throws Exception {
+		final int bufferSize = 1024;
+
+		NetworkBufferPool networkBufferPool = new NetworkBufferPool(10, bufferSize, 2);
+		SingleInputGate inputGate = createSingleInputGate(1);
+		RemoteInputChannel inputChannel = createRemoteInputChannel(inputGate, null, networkBufferPool);
+		inputGate.assignExclusiveSegments();
+
+		CreditBasedPartitionRequestClientHandler handler = new CreditBasedPartitionRequestClientHandler();
+		handler.addInputChannel(inputChannel);
+
+		try {
+			Buffer buffer = TestBufferFactory.createBuffer(bufferSize);
+			BufferResponse bufferResponse = createBufferResponse(
+				buffer,
+				0,
+				inputChannel.getInputChannelId(),
+				1,
+				new NetworkBufferAllocator(handler));
+
+			handler.removeInputChannel(inputChannel);
+			handler.channelRead(null, bufferResponse);
+
+			assertNotNull(bufferResponse.getBuffer());
+			assertTrue(bufferResponse.getBuffer().isRecycled());
+		} finally {
+			releaseResource(inputGate, networkBufferPool);
+		}
+	}
+
+	@Test
+	public void testReceivedBufferForReleasedChannel() throws Exception {
+		final int bufferSize = 1024;
+
+		NetworkBufferPool networkBufferPool = new NetworkBufferPool(10, bufferSize, 2);
+		SingleInputGate inputGate = createSingleInputGate(1);
+		RemoteInputChannel inputChannel = createRemoteInputChannel(inputGate, null, networkBufferPool);
+		inputGate.assignExclusiveSegments();
+
+		CreditBasedPartitionRequestClientHandler handler = new CreditBasedPartitionRequestClientHandler();
+		handler.addInputChannel(inputChannel);
+
+		try {
+			Buffer buffer = TestBufferFactory.createBuffer(bufferSize);
+			BufferResponse bufferResponse = createBufferResponse(
+				buffer,
+				0,
+				inputChannel.getInputChannelId(),
+				1,
+				new NetworkBufferAllocator(handler));
+
+			inputGate.close();
```

Comment:
Can you check whether we already have the case that releasing the channel before `createBufferResponse`, then we can verify whether the created `BufferResponse` has the `null` data buffer.

Reply:
There should not be such tests now, thus new tests are added. The common code is also extracted into utility method.